### PR TITLE
[v1.1][ISSUE-197]DDP grads not sync bug fix in domain adapter

### PR DESCRIPTION
**Pull request by zhouyu5**
### What changes were proposed in this pull request?

- Fix the bug where grads not sync in DDP training of domain adapter
- **The root cause of the issue**:
    - The model is composed of two submodules, classifier network and discriminator network, let's call them model A and model B, each model have their own loss function, say loss A and loss B
    - We used a model T to combine the model A and model B, then wrap model T with `DistributedDataParallel`
    - We want to calculate the total loss, which is `loss A + loss B`, so we take out the model A and model B to perform forward pass, here model A and model B are non-DDP object
    - This brings a hidden problem, when perform backward process, the forward pass and backward pass function is not wrapped by DDP, so end up with n node training their own model without communication
- Many thanks to @Peach-He for helping to point out the problem
- See the fix details in the [last section](#guide)

### Why are the changes needed?

- If not fixing, the ddp training seems like training separately in each node, and users will finally get `n` independent model, this is not what we want to see.

### How was this patch tested?

- After fixing, we see network communication
network-image
- After fixing, we check params' grads after one backwards by using the following code, we saw all params are equal across different node:
```python
for name, params in self.network.named_parameters():
    if params.grad is None: continue
    print(f'rank: {dist.get_rank()}, name: {name}, 
        grads: {torch.sum(params.grad)}, value: {torch.sum(params)}')
```
- Last, we using the exist ci/cd of domain adapter to validate

### <h2 id="guide">Step-by-Step guide on calling DistributedDataParallel on multiple Modules</h2>

- Step 1: We wrap each sub-model in GAN-like model with `DistributedDataParallel`, such that grads of all params are not forgotten in backward pass
```diff
- self.network = DDP(self.network)
+ self.network.backbone = DDP(self.network.backbone)
+ self.network.adapter = DDP(self.network.adapter)
```

- Step 2: If your metric function or loss function is an attribute of sub-model, don't forget to unwrap it before getting its attribute
```diff
-  backbone_loss = self.network.backbone.loss(...)
+  backbone_loss = self.network.backbone.module.loss(...)
```

- Step 3: If you direct save the combined network, you need to remove the `module.` prefix in the `state_dict()`
```python
curr_state_dict_keys = list(self.network.state_dict().keys())
for k, value in checkpoint['state_dict'].items():
      key = k
      if key not in curr_state_dict_keys:
          if 'module.' in k:
              key = k.replace('module.', '')
      new_state_dict[key] = value
```

### Reference
- [What is the best practice for running distributed adversarial training?](https://discuss.pytorch.org/t/what-is-the-best-practice-for-running-distributed-adversarial-training/43323)
- [Calling DistributedDataParallel on multiple Modules?](https://discuss.pytorch.org/t/calling-distributeddataparallel-on-multiple-modules/38055)

### CODE REVIEW

**Review by XinyaoWa — APPROVED** [review]

File: modelzoo/unet/nnunet.patch [hunk 1]
```diff
@@ -29,7 +29,7 @@ diff -urN nnUNet/.github/ISSUE_TEMPLATE/all-issues.md nnUNetNew/.github/ISSUE_TE
 -
 -
 diff -urN nnUNet/.gitignore nnUNetNew/.gitignore
---- nnUNet/.gitignore	2023-02-09 08:27:01.510133303 +0000
+--- nnUNet/.gitignore	2023-02-18 06:51:01.252097744 +0000
 +++ nnUNetNew/.gitignore	1970-01-01 00:00:00.000000000 +0000
```
Comment (xuechendi) [inline modelzoo/unet/nnunet.patch:33]: @zhouyu5 , can we remove the change to .git and .gitignore? Or it will always change

Comment (xuechendi) [reply modelzoo/unet/nnunet.patch:33]: @zhouyu5 , @XinyaoWa , update my request, can we remove unnecessary changes of time in this patch?

File: modelzoo/unet/nnunet.patch [hunk 2]
```diff
@@ -2547,13 +2548,13 @@ diff -urN nnUNet/nnunet/network_architecture/DA/CAC_UNet.py nnUNetNew/nnunet/net
...
 +        source_output, *source_feat = self.backbone(source_data)
-+        source_loss = self.backbone.loss(source_output, source_label)
++        source_loss = self.backbone_loss(source_output, source_label)
 +        
 +        target_loss = None
 +        output, *target_feat = self.backbone(data)
 +        output = [source_pred_to_target_pred(item) for item in output]
 +        if self.target_loss_weight > 0:
-+            target_loss = self.backbone.loss(output, target)
++            target_loss = self.backbone_loss(output, target)
```
Comment (xuechendi) [inline modelzoo/unet/nnunet.patch:2557]: @zhouyu5 , does that mean the sync issue is caused by incorrect function call? what is the difference between `backbone.loss` and `self.backbone.loss`?

**Comment by xuechendi** [comment]
@zhouyu5, can you wrap up your current solution here, including root cause, what is the way you fixed this issue, how this fix being verified, will there be performance impact?

**Comment by xuechendi** [comment]
@zhouyu5 , there are two patch files: modelzoo/unet/nnunet.patch and git.patch
I assume nnunet.patch is not needed anymore? Can you rename git.patch to nnunet.patch?

File: modelzoo/unet/nnunet.patch (hunk 1 above)
Comment (zhouyu5) [reply modelzoo/unet/nnunet.patch]: Already replace the diff patch with git patch. Now no diff on timestamp and .git file.

File: modelzoo/unet/nnunet.patch (hunk 2 above)
Comment (zhouyu5) [reply modelzoo/unet/nnunet.patch:932]: No. I have already update the root cause in PR description. Please have a look~

**Comment by zhouyu5** [comment]
> @zhouyu5 , there are two patch files: modelzoo/unet/nnunet.patch and git.patch I assume nnunet.patch is not needed anymore? Can you rename git.patch to nnunet.patch?

updated.

**Comment by zhouyu5** [comment]
> @zhouyu5, can you wrap up your current solution here, including root cause, what is the way you fixed this issue, how this fix being verified, will there be performance impact?

All updated, including root cause, solution details, verified process.
BTW, there is no impact on performance.

**Comment by zhouyu5** [comment]
@xuechendi Please have a review~It is ready for merge now.

**Review by xuechendi — APPROVED** [review]
LGTM, approved
